[Feat] Add support for group skins from QuakMDL files

yashnator opened the pull request:
### Describe your changes

- Add support for group skins from QuakeMDL format
- Minor fixes and changes to only load the selected skins. Earlier, all the skins were loaded in memory even if they were discarded later

### Issue ticket number and link if any

Fixes #2137

### Checklist for finalizing the PR

- [x] I have performed a [self-review](https://f3d.app/doc/dev/CODING_STYLE.html) of my code
- [x] I have added [tests](https://f3d.app/doc/dev/TESTING.html) for new features and bugfixes
- [x] I have added [documentation](https://f3d.app/) for new features
- [ ] If it is a modifying the libf3d API, I have updated bindings
- [ ] If it is a modifying the `.github/workflows/versions.json`, I have updated `timestamp`

### Continuous integration

Please check the checkbox of the CI you want to run, then push again on your branch.

- [x] Style checks
- [x] Fast CI
- [x] Coverage cached CI
- [x] Analysis cached CI
- [x] WASM docker CI
- [x] Android docker CI
- [x] macOS Intel cached CI
- [x] macOS ARM cached CI
- [x] Windows cached CI
- [x] Linux cached CI
- [x] Other cached CI

mwestphal commented:
Looks good! Ill review when I can!

mwestphal commented on application/testing/CMakeLists.txt:
```diff
@@ -510,7 +510,7 @@ f3d_test(NAME TestVerboseAnimationWrongAnimationTimeLow DATA BoxAnimated.gltf AR
 f3d_test(NAME TestCommandScriptVerboseMultiAnimationTimeRange SCRIPT TestCommandScriptVerboseMultiAnimationTimeRange.txt DATA InterpolationTest.glb ARGS --verbose REGEXP "0, 1.70833" NO_BASELINE)# cycle_animation x3
 f3d_test(NAME TestVerboseQuakeMDLAnimationNoNamingScheme ARGS --verbose DATA v_rock2.mdl REGEXP "0: flame" NO_BASELINE)
-f3d_test(NAME TestVerboseQuakeMDLGroupSkinUnsupported ARGS --verbose DATA groupskin.mdl REGEXP "Groupskin are not supported" NO_BASELINE)
+f3d_test(NAME TestVerboseQuakeMDLGroupSkin ARGS --verbose DATA groupskin.mdl REGEXP "0: group_skin" NO_BASELINE)
```
you will need to remove this test and instead add a standard test, probably similar to `TestQuakeMDLAnimationSimpleFrame`

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -109,45 +109,81 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
+    {
+      int skinGroup = *reinterpret_cast<const int*>(buffer.data() + offset);
+      offset += sizeof(int32_t);
+      // Skip the skins that are not selected
```
I think this comment should be two lines below?

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -109,45 +109,81 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
+        offset += sizeof(int32_t);
+        if (i == skinIndex)
+        {
+          GroupSkins.resize(nb);
```
use `this` everywhere

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -109,45 +109,81 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
+            GroupSkinDurations[j] = *reinterpret_cast<const float*>(buffer.data() + offset);
+            offset += sizeof(float);
+          }
+          // Take cumulative sum of durations for binary search during rendering
```
I'm not sure why this is needed ?

yashnator replied:
I was trying to make the logic similar to the update of vertices in `vtkF3DQuakeMDLImporter::UpdateAtTimeValue`, which uses `std::lower_bound` to find the closest frame

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -428,6 +464,9 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
   std::vector<std::vector<vtkSmartPointer<vtkPolyData>>> AnimationFrames;
 
+  [[maybe_unused]] std::vector<vtkSmartPointer<vtkImageData>> GroupSkins;
+  [[maybe_unused]] std::vector<float> GroupSkinDurations;
```
hum, I dont see why you would need maybe_unused ?

yashnator replied:
We use these two vectors only when we have a group skin selected. Should I remove this?

mwestphal commented:
Hi @yashnator , how is it going ? Do you need a review ?

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -111,45 +112,81 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
+        {
+          unsigned char index = *(buffer.data() + offset + x * skinWidth + y);
+          unsigned char* ptr = static_cast<unsigned char*>(skin->GetScalarPointer(y, x, 0));
+          std::copy(F3DMDLDefaultColorMap[index], F3DMDLDefaultColorMap[index] + 3, ptr);
```
can you copy the whole texture in a single call ? this is not efficient

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -111,45 +112,81 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
-    const unsigned char* selectedSkin = skins[skinIndex].skin;
-    for (int i = 0; i < skinHeight; i++)
+
+    vtkSmartPointer<vtkImageData> img;
```
```suggestion
    vtkNew<vtkImageData> img;
```

mwestphal reviewed (changes requested):
Ive made a first review; make_new_skin seems unefficient :/

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -111,46 +112,77 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
+        this->GroupSkinAnimationNames.emplace_back(skinAnimationName);
+        int nb;
+        std::memcpy(&nb, buffer.data() + offset, sizeof(int));
+        offset += sizeof(int32_t);
```
```suggestion
        offset += sizeof(nb);
```

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -111,46 +112,77 @@ struct vtkF3DQuakeMDLImporter::vtkInternals
+          float timeValue;
+          memcpy(&timeValue, buffer.data() + offset, sizeof(float));
+          this->GroupSkinDurations[groupSkinCount][j] = static_cast<double>(timeValue);
+          offset += sizeof(float);
```
```suggestion
          offset += sizeof(timeValue);
```

mwestphal commented:
Can you rebase/merge with master ? There is a strange error in CI

yashnator commented:
@mwestphal I had rebased onto master before the last force-push. The branch is not behind master, but CI failed :(

mwestphal commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -466,40 +502,63 @@ bool vtkF3DQuakeMDLImporter::UpdateAtTimeValue(double timeValue)
+    }
+    else
+    {
+      this->Internals->Texture->SetInputData(this->Internals->GroupSkins[animIndex][frameIndex]);
```
codecov reports that this code is not tested, can you check ?

yashnator commented on plugins/native/module/vtkF3DQuakeMDLImporter.cxx:
```diff
@@ -466,40 +502,63 @@ bool vtkF3DQuakeMDLImporter::UpdateAtTimeValue(double timeValue)
+      static_cast<vtkIdType>(this->Internals->AnimationNames.size());
+    size_t animIndex = isMeshAnimation
+      ? this->Internals->ActiveAnimation
+      : this->Internals->ActiveAnimation - this->Internals->AnimationNames.size();
```
Ah okay. I see the problem, I should've selected the groupskin for the test now that it is a separate animation

yashnator commented:
> You are modifying versions.json, please update the timestamp as well, this will generate new docker images and caches needed for CI.

Please ignore. I had squashed commits, and it changed hashes. Fixed now

mwestphal commented:
Thanks for the contribution @yashnator !
